gooo/gooo.slide

```diff
@@ -0,0 +1,229 @@
+
+.image ./img/tellme.jpg
+
+
+* That which is in between: Interfaces
+
+
+* Example: Reading File (or Stream...or whatever)
+
+.image ./img/iointeraction.png _ 700
+
+
+* What is in between
+
+.image ./img/iointeraction-between.png _ 650
+
+
+* What types matter ?
+
+.image ./img/iointeraction-types-matter.png _ 750
+
+
+* What Go enables
+
+.image ./img/iointeraction-go-interfaces.png _ 900
+
+
+* What Go enables
+
+.image ./img/iointeraction-go-mul-interfaces.png _ 900
+
+
+* Why is this great ?
+
+- No need for the object implementing an interface to explicit declare it
+- The object implementing an interface is also not coupled to which name has been given to a interface (doesn't matter)
+- The object implementing an interface is also not coupled to a interface type (doesn't matter)
+- For the client, the type (form) of the object implementing an interface is irrelevant
+- The focus is on which messages (methods) the client needs and checking if an object has them
+
+
+* Boils down to responsibilities
+
+- Client just needs to know what it wants, and define an interface
+- Service just needs to have methods
+- The compiler does the job of checking if the service is compatible with the interface declared by the client
+- Low coupling =)
+
+
+* Flexible and Safe
+
+There is no need to design ahead of time which interfaces you
+are going to need.
+
+No need to previously extrapolate how your objects will be used.
+
+Interfaces can be added to communicate with pre-existent objects as
+the need arises, in ways you did not foreseen, safely.
+
+Changes on a object that breaks a interface is caught
+at compile time with a clear error. No runtime surprises.
+
+Basically the flexibility of a dynamic typed language
+with compile time safety of a static one.
+
+
+* Real life example
+
+
+* How to get the status code of a exec.Command ?
+
+
+* Seems trivial...
+
+- Actually not
+- Some operational systems uses integers
+- Others have strings as the results of a process execution
+- Officially exported API on exec.Command don't give access to the status code (only if it failed or not)
+
+
+* Investigating the error
+
+.code ./code/exiterror.go
+
+
+* Investigating the ProcessState
+
+.code ./code/processstate.go
+
+
+* Empty interfaces...
+
+
+// TODO: some funny meme
+
+
+* Checking the unix implementation (WaitStatus)
+
+.code ./code/waitstatus.go
+
+
+* Found it
+
+// TODO: some other funny meme about finding ?
+
+
+* What about other platforms ?
+
+
+	syscall % grep -R ExitStatus .
+	./syscall_nacl.go:func (w WaitStatus) ExitStatus() int    { return 0 }
+	./syscall_bsd.go:func (w WaitStatus) ExitStatus() int {
+	./syscall_solaris.go:func (w WaitStatus) ExitStatus() int {
+	./syscall_linux.go:func (w WaitStatus) ExitStatus() int {
+	./syscall_windows.go:func (w WaitStatus) ExitStatus() int { return int(w.ExitCode) }
+	./syscall_plan9.go:func (w Waitmsg) ExitStatus() int { 
+
+
+* Its Something
```

Comment:
It's

Reply:
Im lazy when it comes to reaching out for ' =P